src/get.rs

```diff
@@ -129,14 +134,171 @@ impl AsyncRead for DataStream {
     }
 }
 
+/// Gets a collection and all its blobs using a [`Ticket`].
+pub async fn run_ticket<A, B, C, FutA, FutB, FutC>(
+    ticket: Ticket,
+    keylog: bool,
+    max_concurrent: u8,
+    on_connected: A,
+    on_collection: B,
+    on_blob: C,
+) -> Result<Stats>
+where
+    A: FnOnce() -> FutA,
+    FutA: Future<Output = Result<()>>,
+    B: FnOnce(&Collection) -> FutB,
+    FutB: Future<Output = Result<()>>,
+    C: FnMut(Hash, DataStream, String) -> FutC,
+    FutC: Future<Output = Result<DataStream>>,
+{
+    let start = Instant::now();
+    let connection = dial_ticket(&ticket, keylog, max_concurrent.into()).await?;
+    run_connection(
+        connection,
+        ticket.hash,
+        ticket.token,
+        start,
+        on_connected,
+        on_collection,
+        on_blob,
+    )
+    .await
+}
+
+async fn dial_ticket(
+    ticket: &Ticket,
+    keylog: bool,
+    max_concurrent: usize,
+) -> Result<quinn::Connection> {
+    // Sort the interfaces to make sure local ones are at the front of the list.
+    let interfaces = default_net::get_interfaces();
+    let mut addrs: Vec<SocketAddr> = ticket
+        .addrs
+        .iter()
+        .filter(|addr| is_same_subnet(addr, &interfaces))
+        .copied()
+        .collect();
+    let other_addrs: Vec<SocketAddr> = ticket
+        .addrs
+        .iter()
+        .filter(|addr| !addrs.contains(addr))
+        .copied()
+        .collect();
+    addrs.extend(other_addrs);
+    let cancel_token = CancellationToken::new();
+    let (results_tx, mut results_rx) = mpsc::unbounded_channel();
+    tokio::spawn(dial_all_task(
+        addrs,
+        ticket.peer,
+        keylog,
+        results_tx,
+        max_concurrent,
+        cancel_token.clone(),
+    ));
+    loop {
+        match results_rx.recv().await {
+            Some(Ok(conn)) => {
+                cancel_token.cancel();
+                return Ok(conn);
+            }
+            Some(Err(_)) => (),
+            None => bail!("Failed to establish connection to peer"),
+        }
+    }
+}
+
+fn is_same_subnet(addr: &SocketAddr, interfaces: &[Interface]) -> bool {
+    for interface in interfaces {
+        match addr {
+            SocketAddr::V4(peer_addr) => {
+                for net in interface.ipv4.iter() {
+                    if same_subnet_v4(net.addr, *peer_addr.ip(), net.prefix_len) {
+                        return true;
+                    }
+                }
+            }
+            SocketAddr::V6(peer_addr) => {
+                for net in interface.ipv6.iter() {
+                    if same_subnet_v6(net.addr, *peer_addr.ip(), net.prefix_len) {
+                        return true;
+                    }
+                }
+            }
+        }
+    }
+    false
+}
+
+/// Task which will dial all addresses concurrently.
+///
+/// Results are sent back to `results_tx`.  Dialing can be interrupted by cancelling the
+/// `cancel_token`.
+async fn dial_all_task(
```

Comment:
why manually implement this, and not sth like https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffer_unordered

Reply:
huh, that is a terribly named method.  but it is a joy to use!  TIL, I was indeed surprised how much work this was.